rpc/subscriptions.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/jsonrpc"
+)
+
+const subscribeEventsChunkSize = 1024
+
+type SubscriptionResponse struct {
+	Version string `json:"jsonrpc"`
+	Method  string `json:"method"`
+	Params  any    `json:"params"`
+}
+
+func (h *Handler) SubscribeEvents(ctx context.Context, fromAddr *felt.Felt, keys [][]felt.Felt,
+	blockID *BlockID,
+) (*SubscriptionID, *jsonrpc.Error) {
+	w, ok := jsonrpc.ConnFromContext(ctx)
+	if !ok {
+		return nil, jsonrpc.Err(jsonrpc.MethodNotFound, nil)
+	}
+
+	lenKeys := len(keys)
+	for _, k := range keys {
+		lenKeys += len(k)
+	}
+	if lenKeys > maxEventFilterKeys {
+		return nil, ErrTooManyKeysInFilter
+	}
+
+	var requestedHeader *core.Header
+	headHeader, err := h.bcReader.HeadsHeader()
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err.Error())
+	}
+
+	if blockID == nil {
+		requestedHeader = headHeader
+	} else {
+		if blockID.Pending {
+			return nil, ErrCallOnPending
+		}
+
+		var rpcErr *jsonrpc.Error
+		requestedHeader, rpcErr = h.blockHeaderByID(blockID)
+		if rpcErr != nil {
+			return nil, rpcErr
+		}
+
+		if headHeader.Number >= maxBlocksBack && requestedHeader.Number <= headHeader.Number-maxBlocksBack {
+			return nil, ErrTooManyBlocksBack
+		}
+	}
+
+	id := h.idgen()
+	subscriptionCtx, subscriptionCtxCancel := context.WithCancel(ctx)
+	sub := &subscription{
+		cancel: subscriptionCtxCancel,
+		conn:   w,
+	}
+	h.mu.Lock()
+	h.subscriptions[id] = sub
+	h.mu.Unlock()
+
+	headerSub := h.newHeads.Subscribe()
+	sub.wg.Go(func() {
+		defer func() {
+			h.unsubscribe(sub, id)
+			headerSub.Unsubscribe()
+		}()
+
+		// The specification doesn't enforce ordering of events therefore events from new blocks can be sent before
+		// old blocks.
+		// Todo: see if sub's wg can be used?
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case <-subscriptionCtx.Done():
+					return
+				case header := <-headerSub.Recv():
+
+					h.processEvents(subscriptionCtx, w, id, header.Number, header.Number, fromAddr, keys)
+				}
+			}
+		}()
+
+		h.processEvents(subscriptionCtx, w, id, requestedHeader.Number, headHeader.Number, fromAddr, keys)
+
+		wg.Wait()
+	})
+
+	return &SubscriptionID{ID: id}, nil
+}
+
+func (h *Handler) processEvents(ctx context.Context, w jsonrpc.Conn, id, from, to uint64, fromAddr *felt.Felt, keys [][]felt.Felt) {
+	filter, err := h.bcReader.EventFilter(fromAddr, keys)
+	if err != nil {
+		h.log.Warnw("Error creating event filter", "err", err)
+		return
+	}
+
+	defer func() {
+		h.callAndLogErr(filter.Close, "Error closing event filter in events subscription")
+	}()
+
+	if err = setEventFilterRange(filter, &BlockID{Number: from}, &BlockID{Number: to}, to); err != nil {
+		h.log.Warnw("Error setting event filter range", "err", err)
+		return
+	}
+
+	filteredEvents, cToken, err := filter.Events(nil, subscribeEventsChunkSize)
```

Comment:
How about `nextToken` istead of `cToken`? - I'd like to make the name more descriptive.

Reply:
`c` is short for continuation, same convention is used in other event related functions.